### PR TITLE
Compress Cloudformation template if it exceeds max allowed length

### PR DESCRIPTION
**Pull request by theJWong**
## Ready State
**Ready**

## Synopsis
Compress Cloudformation template if it exceeds max allowed length.

Compression in this case is to remove whitespaces from the template. The original template remains intact and it is only the rendered template that will be sent to Cloudformation that is compressed. This is to avoid the scenario where our cloudformation template exceeds the length of max allowed template size to be sent over in the API request.

## Changelog
  * Remove whitespaces if the Cloudformation template exceeds the max allowed length / size of the template before sending it to Cloudformation.

## Testing
```
000377-ml:ellation_formation jwong$ ef-cf cloudformation/services/templates/subs.json proto0 --devel
=== DRY RUN ===
Validation only. Use --commit to push template to CF
=== DRY RUN ===
not refreshing repo because --devel was set or running on Jenkins
Template exceeds the max allowed length that Cloudformation will accept. Compressing template...
Uncompressed size of template: 51607
Compressed size of template: 37287
Template passed validation
```

### CODE REVIEW

File: src/ef-cf.py [hunk 1]
```diff
@@ -231,6 +235,14 @@ def main():
   else:
     parameters = []
 
+  # Detect if the template exceeds the maximum size that is allowed by Cloudformation
+  if len(template) > CLOUDFORMATION_SIZE_LIMIT:
```
Comment (bobsut) [inline src/ef-cf.py]: Is there any reason not to space-compress all whitespace in every template, regardless of the original length?
You could remove the conditional and still get the early warning that you're nearing the max length
```
print("Compressed from {} ".format(len(template)), end='')
template = template.replace("\s+", "")
print("to {} bytes of 51200 allowed".format(len(template)))
```

Comment (theJWong) [reply src/ef-cf.py]: I rather not perform the compression unless I absolutely have to. I foresee some problems with completely removing whitespaces including metadata or descriptions inside of the cloudformation templates that will get removed from this pseudo compression.

File: src/ef-cf.py [hunk 2]
```diff
@@ -231,6 +235,14 @@ def main():
...
+    # Compress the generated template by removing whitespaces
+    print("Template exceeds the max allowed length that Cloudformation will accept. Compressing template...")
+    print("Uncompressed size of template: {}".format(len(template)))
+    template = template.replace(" ", "")
```
Comment (momoneko) [inline src/ef-cf.py]: Removing the spaces is kind of unsafe, you will change the data in the strings as well. 

This is safer, and looks like it compresses it even more than just removing the spaces
```  
    unpacked = json.loads(template)
    template = json.dumps(unpacked, indent=0, separators=(",",":")).replace('\n', '')
```

Comment (theJWong) [reply src/ef-cf.py]: Perfect will do that and run additional tests.

Comment (momoneko) [reply src/ef-cf.py]: @theJWong, small update; it should be enough to only do 
`json.dumps(unpacked,  separators=(",", ":"))`. No need for the replace part.

**Pull request merged**